tripal/tripal
# Provides Page Title Format validation for both UI and publish

# New Feature

### Closes #2009

### Depends on #2040 (merged)

## Description
This adds validation of the page title format to both UI entry of a new entity, and to publish.
Two error conditions are checked:
1. No tokens in the title format - this would cause all titles to be the same
2. Title format is still the generic default used when a new content type is created through the UI, i.e.
 `Entity [TripalEntity__entity_id]`
However I am not sure how to disable saving of the entity entered through the UI, currently you can just ignore the error message and save anyway.

## Testing?
1. Create a new content type "Test Project", I used term "example of usage (IAO:0000112)".
2. Note the default title format, don't change it yet.
![Issue2009 1](https://github.com/user-attachments/assets/424aa273-a6f6-4c15-8031-c0468761284e)
4. Add field "Test Project Name" of a "Chado String Field Type", select "project" table, column "name", term "name (schema:name)".
5. Add field "Test Project Description" of a "Chado Text Field Type", column "description", term "description (schema:description)".
6. Start to create a new entity of this content type. You should get an error message.
![Issue2009 2](https://github.com/user-attachments/assets/2ac1f2e1-1745-4398-90f8-60903cae2c18)

7. Click on the link in the warning to confirm it goes to the right page.
8. Try to publish this content type. Insert a record
 `INSERT INTO project (name, description) VALUES ('PublishTest', 'RandomDescription')`
9. Publish through GUI and run the job, or use command line `drush tripal-chado:publish test_project` You should get the same error both ways.
![Issue2009 3](https://github.com/user-attachments/assets/6ae9127c-52ce-4199-adc3-4554081fa4d3)

10. Update the title format to something without any tokens, this should be caught.
![Issue2009 4](https://github.com/user-attachments/assets/fec765b3-b239-44bc-8ce0-430416463b83)
![Issue2009 5](https://github.com/user-attachments/assets/03c88e73-2240-4b89-8eff-17639850b964)

11. Update the title format to e.g. `[field_test_project_name]`. You should now be able to both create through the UI and publish.
However I think issue #2032 is happening here too, commit https://github.com/tripal/tripal/pull/2052/commits/1f966b76a9ac2051f33d16f8184cbd28fb96a8ff was a workaround.
![Issue2009 6](https://github.com/user-attachments/assets/e2b68271-ca76-4592-9e0e-5cb44c6cc0d8)
